[Merged by Bors] - refactor(CategoryTheory/Sums): refactor sums of categories

robin-carlier opened the pull request:
We refactor the category instance on sums of types so that it is defined even for types living in a different universes.

Consequently, we change the order of things related to sums of categories:
  - We remove `simp` instance on `Sum.inl_.map` and `Sum.inr_.map`: the inclusions for maps should be considered primitives.
  - We make `Sum.sum'` the first definition about functors out of a sum.
  - We add an extensionality (up to isomorphism) theorem for functors out of sums in terms of their precomposition with `inl_`and `inr_`.
  - We redefine `Sum.swap`, `Sum.associator` (as well as its inverse), `Functor.sum` to use only `Sum.sum'`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

This refactor follows discussions in #22278 and #22281 and is rather opinionated: the universal property of sums of categories should be the only way one should work with sums and should be the only way one defines functors out of sums.

The proof of the associativity equivalence of categories becomes a bit longer, but I believe it is the morally right one.

This was an occasion to clean up erw's as well.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

b-mehta commented:
I'm not convinced the universe generalisation is worth it, particularly so because it makes the description of homs in the new category awkward: we can just ulift the two input categories from v_1 and v_2 to max v_1 v_2 and take the sum there, now that they match. The other changes you get from this are nice though...

robin-carlier commented:
To me, I believe the "akwardness" of the hom is OK if we consider that homs in a sum of categories should be treated as "private" that can only be accessed through `inl_` and `inr_`.
Perhaps I could add the following so that it still remains "easy" to work with these homs?
```lean
@[elab_as_elim, cases_eliminator, induction_eliminator]
def homInduction {P : {x y : C ⊕ D} → (x ⟶ y) → Sort*} 
    (inl : ∀ x y : C, (f : x ⟶ y) → P ((inl_ C D).map f)) 
    (inr : ∀ x y : D, (f : x ⟶ y) → P ((inr_ C D).map f))
    {x y : C ⊕ D} (f : x ⟶ y) : P f :=
  match x, y, f with
  | .inl x, .inl y, f => inl x y f.down
  | .inr x, .inr y, f => inr x y f.down

@[simp]
def homInduction_left {P : {x y : C ⊕ D} → (x ⟶ y) → Sort*} 
    (inl : ∀ x y : C, (f : x ⟶ y) → P ((inl_ C D).map f)) 
    (inr : ∀ x y : D, (f : x ⟶ y) → P ((inr_ C D).map f))
    {x y : C} (f : x ⟶ y) : homInduction (P := P) _ _ inl inr ((inl_ C D).map f) = inl x y f :=
  rfl

@[simp]
def homInduction_right {P : {x y : C ⊕ D} → (x ⟶ y) → Sort*} 
    (inl : ∀ x y : C, (f : x ⟶ y) → P ((inl_ C D).map f)) 
    (inr : ∀ x y : D, (f : x ⟶ y) → P ((inr_ C D).map f))
    {x y : D} (f : x ⟶ y) : homInduction (P := P) _ _ inl inr ((inr_ C D).map f) = inr x y f :=
  rfl
```

robin-carlier commented:
I added that induction principle and changed a bit the simp lemmas: it looks like this plays very nicely, and that now everything except for `homInduction` does not refer to `ULift.down`, so that the underlying nature of the hom types is completely hidden from the user.

b-mehta commented:
Ah, the change in the simp lemmas makes this a lot more convincing; nice stuff!

kim-em commented on Mathlib/CategoryTheory/Sums/Basic.lean:
```diff
@@ -203,15 +196,69 @@ theorem sum_obj_inr (F : A ⥤ B) (G : C ⥤ D) (c : C) : (F.sum G).obj (inr c)
+section
+
+variable {F G: A ⊕ B ⥤ C}
+  (e₁ : Sum.inl_ _ _ ⋙ F ≅ Sum.inl_ _ _ ⋙ G)
+  (e₂ : Sum.inr_ _ _ ⋙ F ≅ Sum.inr_ _ _ ⋙ G)
```
I think it's good style to include the explicit variables in each `def` (but not `theorem`), for readability. In this case one doesn't have to look far, but still!

kim-em commented on Mathlib/CategoryTheory/Sums/Basic.lean:
```diff
@@ -203,15 +196,69 @@ theorem sum_obj_inr (F : A ⥤ B) (G : C ⥤ D) (c : C) : (F.sum G).obj (inr c)
+
+section
+
+variable (F : A ⊕ B ⥤ C) (a : A) (b : B)
```
Similar style point: I'd actually move the `(a : A) (b : B)` below the `def`, so no human elaboration is required to know that they are not used by the `def`.

kim-em commented:
bors d+

robin-carlier commented:
bors r+
